File: app/src/main/java/com/example/services/LifeService.kt
package com.example.services

import com.example.data.local.entity.LifeEntity
import com.example.data.model.ClientCharacter
import com.example.data.repository.DestinyRepository
import com.example.data.repository.LifeRecordSummary
import kotlinx.coroutines.flow.Flow

/**
 * LIFE SERVICE (Prompt 01 of 20 Foundation)
 *
 * Core service managing:
 * 1. Life creation lifecycle (User -> SaveSlot -> Life -> Character)
 * 2. NPC Simulation Model:
 *    - "dormant": NPC is lightweight. Attributes and stats persist in the database,
 *      but no annual simulation logic, relationship updates, or event checks are executed.
 *    - "active": NPC has entered player interaction horizon. The simulation engine
 *      actively calculates events, momentum shifts, and decisions.
 * 3. Safe client encapsulation (Karma privacy enforcement).
 */
class LifeService(private val repository: DestinyRepository) {

    val allLives: Flow<List<LifeEntity>> = repository.allLives
    val activeLives: Flow<List<LifeEntity>> = repository.activeLives
    val dormantLives: Flow<List<LifeEntity>> = repository.dormantLives

    suspend fun createNewLife(
        playerName: String,
        saveSlotName: String,
        characterName: String,
        birthYear: Int = 2000,
        isDormant: Boolean = false,
        birthCountry: String = "United States",
        birthCity: String = "New York",
        gender: String = "Male",
        sexuality: String = "Heterosexual",
        talent: String = "None",
        eyeStyle: String = "Almond",
        eyeColor: String = "Brown",
        skinTone: String = "Fair",
        browStyle: String = "Straight",
        facialHairStyle: String = "Clean Shaven",
        facialHairColor: String = "Black",
        hairStyle: String = "Short Crop",
        hairColor: String = "Black",
        intelligence: Int? = null,
        discipline: Int? = null,
        willpower: Int? = null,
        ambition: Int? = null,
        health: Int? = null,
        looks: Int? = null,
        smarts: Int? = null,
        happiness: Int? = null,
        fertility: Int? = null,
        energy: Int? = null,
        athleticPerformance: Int? = null
    ): LifeRecordSummary {
        return repository.createNewLife(
            playerName = playerName,
            saveSlotName = saveSlotName,
            characterName = characterName,
            birthYear = birthYear,
            isDormant = isDormant,
            birthCountry = birthCountry,
            birthCity = birthCity,
            gender = gender,
            sexuality = sexuality,
            talent = talent,
            eyeStyle = eyeStyle,
            eyeColor = eyeColor,
            skinTone = skinTone,
            browStyle = browStyle,
            facialHairStyle = facialHairStyle,
            facialHairColor = facialHairColor,
            hairStyle = hairStyle,
            hairColor = hairColor,
            intelligence = intelligence ?: (45..90).random(),
            discipline = discipline ?: (40..85).random(),
            willpower = willpower ?: (40..85).random(),
            ambition = ambition ?: (45..90).random(),
            health = health ?: (60..98).random(),
            looks = looks ?: (40..90).random(),
            smarts = smarts ?: (45..90).random(),
            happiness = happiness ?: (50..90).random(),
            fertility = fertility ?: (70..95).random(),
            energy = energy ?: (80..100).random(),
            athleticPerformance = athleticPerformance ?: (40..85).random()
        )
    }

    suspend fun updateCharacterAgingStats(
        characterId: String,
        health: Int,
        fertility: Int,
        energy: Int,
        athleticPerformance: Int
    ) {
        repository.updateCharacterAgingStats(characterId, health, fertility, energy, athleticPerformance)
    }

    suspend fun updateLifeAge(lifeId: String, newAge: Int) {
        repository.updateLifeAge(lifeId, newAge)
    }

    suspend fun getLifeDetails(lifeId: String): LifeRecordSummary? {
        return repository.getLifeSummary(lifeId)
    }

    /**
     * Transition NPC between Dormant and Active states.
     * Architectural guarantee: Switching is lightweight and zero-data-loss.
     */
    suspend fun setNpcSimulationState(lifeId: String, isDormant: Boolean) {
        repository.setLifeDormancy(lifeId, isDormant)
    }

    /**
     * Tests a resolution check for a character using the shared FormulaEngine.
     * Extracts intrinsic foundation stats from the client character and runs through the 5-layer pipeline.
     */
    fun evaluateCharacterProspect(
        character: ClientCharacter,
        trajectoryMomentum: Double = 0.0,
        luckRoll: Double? = null
    ): FormulaEngine.Outcome {
        val foundation = FormulaEngine.CharacterFoundation(
            intelligence = character.intelligence.toDouble(),
            discipline = character.discipline.toDouble(),
            willpower = character.willpower.toDouble(),
            ambition = character.ambition.toDouble(),
            health = character.health.toDouble(),
            looks = character.looks.toDouble(),
            smarts = character.smarts.toDouble(),
            happiness = character.happiness.toDouble()
        )

        val momentum = FormulaEngine.Momentum(
            trajectoryScore = trajectoryMomentum,
            recentStreak = 0,
            fatigueFactor = 0.0
        )

        val luck = luckRoll?.let { FormulaEngine.Luck(rawRoll = it) } ?: FormulaEngine.Luck()

        return FormulaEngine.evaluate(
            characterFoundation = foundation,
            momentum = momentum,
            luck = luck
        )
    }
}

File: services/lifeService.ts
/**
 * LIFE SERVICE & FIRESTORE PERSISTENCE ORCHESTRATOR
 * Master Build Prompts 01-03: FOUNDATION, CHARACTER SYSTEM, FAMILY SYSTEM
 *
 * Core service managing:
 * 1. Life creation pipeline (User -> SaveSlot -> Life -> Character) in Firestore
 * 2. Unlimited save slots per user
 * 3. NPC Simulation Model:
 *    - 'dormant': Lightweight. Stats stored in DB, but not actively ticked year-to-year.
 *    - 'active': Full simulation active because player is currently interacting.
 * 4. Karma Privacy Mandate: Strips karma from all client-facing responses.
 */

import {
  doc,
  getDoc,
  setDoc,
  updateDoc,
  collection,
  query,
  where,
  getDocs,
  serverTimestamp,
} from 'firebase/firestore';
import { db } from '../lib/firebase';
import {
  Character,
  ClientCharacter,
  CreateLifeInput,
  Life,
  LifeSummaryResponse,
  SaveSlot,
  User,
} from '../types';

export class LifeService {
  /**
   * Sanitizes a Character entity into a ClientCharacter,
   * strictly omitting the internal karma stat.
   */
  public static toClientCharacter(character: Character): ClientCharacter {
    const { karma, ...clientSafe } = character;
    return clientSafe;
  }

  /**
   * Atomic creation flow in Firestore:
   * User → SaveSlot → Life → Character
   */
  public static async createNewLife(input: CreateLifeInput): Promise<LifeSummaryResponse> {
    const {
      username,
      slotName,
      characterName,
      birthYear = 2000,
      isDormant = false,
      birthCountry = 'United States',
      birthCity = 'New York',
      gender = 'Male',
      sexuality = 'Heterosexual',
      talent = 'None',
      eyeStyle = 'Almond',
      eyeColor = 'Brown',
      skinTone = 'Fair',
      browStyle = 'Straight',
      facialHairStyle = 'Clean Shaven',
      facialHairColor = 'Black',
      hairStyle = 'Short Crop',
      hairColor = 'Black',
      intelligence = randomStat(45, 95),
      discipline = randomStat(40, 90),
      willpower = randomStat(40, 90),
      ambition = randomStat(45, 95),
      health = randomStat(60, 99),
      looks = randomStat(40, 90),
      smarts = randomStat(45, 95),
      happiness = randomStat(50, 90),
      fertility = randomStat(70, 95),
      energy = randomStat(80, 100),
      athleticPerformance = randomStat(40, 85),
      geneticHealthModifier = null,
      geneticIntelligenceModifier = null,
      geneticLooksModifier = null,
    } = input;

    const userId = `user_${username.toLowerCase().replace(/[^a-z0-9]/g, '_')}`;
    const userRef = doc(db, 'users', userId);

    // 1. Ensure User entity
    const userSnap = await getDoc(userRef);
    let userData: User;
    const now = new Date();

    if (userSnap.exists()) {
      userData = userSnap.data() as User;
    } else {
      userData = {
        id: userId,
        username,
        email: `${username.toLowerCase()}@destiny.game`,
        createdAt: now,
        updatedAt: now,
      };
      await setDoc(userRef, {
        ...userData,
        createdAt: serverTimestamp(),
        updatedAt: serverTimestamp(),
      });
    }

    // 2. SaveSlot Entity
    const saveSlotId = `slot_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`;
    const saveSlotRef = doc(db, 'saveSlots', saveSlotId);
    const saveSlotData: SaveSlot = {
      id: saveSlotId,
      userId: userData.id,
      slotName: slotName || `Save Slot ${Date.now()}`,
      createdAt: now,
      updatedAt: now,
    };
    await setDoc(saveSlotRef, {
      ...saveSlotData,
      createdAt: serverTimestamp(),
      updatedAt: serverTimestamp(),
    });

    // 3. Life Entity
    const lifeId = `life_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`;
    const lifeRef = doc(db, 'lives', lifeId);
    const lifeData: Life = {
      id: lifeId,
      saveSlotId: saveSlotData.id,
      name: characterName,
      birthYear,
      currentAge: 0,
      isDormant,
      createdAt: now,
      updatedAt: now,
    };
    await setDoc(lifeRef, {
      ...lifeData,
      userId: userData.id,
      createdAt: serverTimestamp(),
      updatedAt: serverTimestamp(),
    });

    // 4. Character Entity
    const characterId = `char_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`;
    const charRef = doc(db, 'characters', characterId);
    const characterData: Character = {
      id: characterId,
      lifeId: lifeData.id,
      intelligence,
      discipline,
      willpower,
      ambition,
      health,
      looks,
      smarts,
      happiness,
      fertility,
      energy,
      athleticPerformance,
      gender,
      sexuality,
      talent,
      eyeStyle,
      eyeColor,
      skinTone,
      browStyle,
      facialHairStyle,
      facialHairColor,
      hairStyle,
      hairColor,
      geneticHealthModifier,
      geneticIntelligenceModifier,
      geneticLooksModifier,
      birthCity,
      birthCountry,
      bankBalance: 1000, // Starting baseline bank balance
      karma: randomStat(30, 80), // HIDDEN server-side
      createdAt: now,
      updatedAt: now,
    };

    await setDoc(charRef, {
      ...characterData,
      userId: userData.id,
      createdAt: serverTimestamp(),
      updatedAt: serverTimestamp(),
    });

    return {
      user: userData,
      saveSlot: saveSlotData,
      life: lifeData,
      character: this.toClientCharacter(characterData),
    };
  }

  /**
   * Updates character stats (e.g. from AgingService or Family events).
   */
  public static async updateCharacterStats(
    characterId: string,
    updates: Partial<Character>
  ): Promise<ClientCharacter> {
    const { karma, ...allowedUpdates } = updates;
    const charRef = doc(db, 'characters', characterId);
    await updateDoc(charRef, {
      ...allowedUpdates,
      updatedAt: serverTimestamp(),
    });

    const updatedSnap = await getDoc(charRef);
    const character = updatedSnap.data() as Character;
    return this.toClientCharacter(character);
  }

  /**
   * Updates life age.
   */
  public static async updateLifeAge(lifeId: string, currentAge: number): Promise<void> {
    const lifeRef = doc(db, 'lives', lifeId);
    await updateDoc(lifeRef, {
      currentAge,
      updatedAt: serverTimestamp(),
    });
  }

  /**
   * Retrieves life details with karma stripped.
   */
  public static async getLifeById(lifeId: string): Promise<LifeSummaryResponse | null> {
    const lifeRef = doc(db, 'lives', lifeId);
    const lifeSnap = await getDoc(lifeRef);
    if (!lifeSnap.exists()) return null;
    const life = lifeSnap.data() as Life;

    // Fetch saveSlot
    const slotRef = doc(db, 'saveSlots', life.saveSlotId);
    const slotSnap = await getDoc(slotRef);
    const saveSlot = slotSnap.exists()
      ? (slotSnap.data() as SaveSlot)
      : ({ id: life.saveSlotId, userId: '', slotName: 'Default', createdAt: new Date(), updatedAt: new Date() });

    // Fetch user
    const userRef = doc(db, 'users', saveSlot.userId);
    const userSnap = await getDoc(userRef);
    const user = userSnap.exists()
      ? (userSnap.data() as User)
      : ({ id: saveSlot.userId, username: 'Player', createdAt: new Date(), updatedAt: new Date() });

    // Fetch character by lifeId
    const charQ = query(collection(db, 'characters'), where('lifeId', '==', lifeId));
    const charDocs = await getDocs(charQ);
    if (charDocs.empty) return null;

    const charDoc = charDocs.docs[0];
    const character = charDoc.data() as Character;

    return {
      user,
      saveSlot,
      life,
      character: this.toClientCharacter(character),
    };
  }

  /**
   * Transitions NPC between Dormant and Active simulation states.
   */
  public static async setNpcDormancy(lifeId: string, isDormant: boolean) {
    const lifeRef = doc(db, 'lives', lifeId);
    await updateDoc(lifeRef, {
      isDormant,
      updatedAt: serverTimestamp(),
    });

    // Also update associated character isDormant if exists
    const charQ = query(collection(db, 'characters'), where('lifeId', '==', lifeId));
    const charDocs = await getDocs(charQ);
    for (const docItem of charDocs.docs) {
      await updateDoc(docItem.ref, {
        isDormant,
        updatedAt: serverTimestamp(),
      });
    }
  }

  /**
   * Resumes any life at any saved age.
   */
  public static async resumeLife(lifeId: string) {
    return this.getLifeById(lifeId);
  }
}

function randomStat(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}
